### PR TITLE
Rooms: Add a hidden |metadata| message for bots

### PR DESCRIPTION
**Pull request by sparkychildcharlie**
- ``|privacy|[secret | hidden | (public)]`` - a display for bots to parse to automatically mark rooms as "private"

### CODE REVIEW

**Comment by sparkychildcharlie** [comment]
In tandem with https://github.com/Zarel/Pokemon-Showdown-Client/pull/913

**Comment by panpawn** [comment]
... What?  This is completely unneeded;  bots can already tell if a room is private or not by doing ``/cmd rooms``.  If the room is not on that list, then the room is private.  That's simple enough logic for a bot programmer to figure out.

**Comment by panpawn** [comment]
In terms of if the room is hidden/secret, I can't really think of any instances where that information would be helpful.  If a room is private in general, it probably shouldn't be leaked.

**Comment by Zarel** [comment]
It's useful information for clients in general, I suppose. But maybe more in a `|metadata|` message rather than having a field for privacy explicitly?

**Comment by Zarel** [comment]
@sparkychildcharlie?

**Comment by Zarel** [comment]
...@sparkychildcharlie?

I'm closing this in another week of inactivity.

**Comment by sparkychildcharlie** [comment]
Uh sorry, I still plan to work on this, but I've been too busy with (prepping for PSPL) IRL stuff.

File: rooms.js [hunk 1]
```diff
@@ -261,6 +261,11 @@ class Room {
 			}
 		}
 	}
+
+	get metadata() {
+		let settings = {title: this.title, id: this.id, privacy: (this.isPrivate === true ? 'secret' : this.isPrivate === 'hidden' ? 'hidden' : 'public'), type: this.type, modchat: this.modchat, modjoin: this.modjoin, auth: (this.auth || {}), uno: (!this.unoDisabled), tournaments: (this.toursEnabled === '%' ? '%' : this.toursEnabled ? '@' : '#'), hangman: (!this.hangmanDisabled)};
```
Comment (Zarel) [inline rooms.js]: This should not be a one-liner

Comment (Zarel) [inline rooms.js]: `const`

File: rooms.js [hunk 2]
```diff
@@ -960,7 +965,7 @@ class BattleRoom extends Room {
 		}
 	}
 	onConnect(user, connection) {
-		this.sendUser(connection, '|init|battle\n|title|' + this.title + '\n' + this.getLogForUser(user).join('\n'));
+		this.sendUser(connection, '|init|battle\n' + this.metadata + '|title|' + this.title + '\n' + this.getLogForUser(user).join('\n'));
```
Comment (Zarel) [inline rooms.js]: metadata should probably go after title

File: rooms.js (hunk 1 above)
Comment (Zarel) [inline rooms.js]: `title` is redundant because `title` is already sent in the Join message.

File: rooms.js [hunk 3]
```diff
@@ -263,7 +263,19 @@ class Room {
 	}
 
 	get metadata() {
-		let settings = {title: this.title, id: this.id, privacy: (this.isPrivate === true ? 'secret' : this.isPrivate === 'hidden' ? 'hidden' : 'public'), type: this.type, modchat: this.modchat, modjoin: this.modjoin, auth: (this.auth || {}), uno: (!this.unoDisabled), tournaments: (this.toursEnabled === '%' ? '%' : this.toursEnabled ? '@' : '#'), hangman: (!this.hangmanDisabled)};
+		let settings = {
+			// settings
+			privacy: (this.isPrivate === true ? 'secret' : this.isPrivate === 'hidden' ? 'hidden' : 'public'),
+			modchat: this.modchat,
+			modjoin: this.modjoin,
+			auth: (this.auth || {}),
```
Comment (Zarel) [inline rooms.js]: ummmm

Comment (Zarel) [inline rooms.js]: This is way too big of a thing to send automatically to clients who didn't ask for it.

File: rooms.js [hunk 4]
```diff
@@ -261,6 +261,22 @@ class Room {
 			}
 		}
 	}
+
+	get metadata() {
+		let settings = {
```
Comment (Zarel) [inline rooms.js]: `const`

File: rooms.js [hunk 5]
```diff
@@ -261,6 +261,22 @@ class Room {
...
+			// settings
+			privacy: (this.isPrivate === true ? 'secret' : this.isPrivate === 'hidden' ? 'hidden' : 'public'),
+			modchat: this.modchat,
+			modjoin: this.modjoin,
+
+			// games
```
Comment (Zarel) [inline rooms.js]: This is still too much metadata to be given automatically. `roommeta` maybe.

Comment (sparkychildcharlie) [inline rooms.js]: Isnt this just an extra ~150 characters

``|metadata|{"title":"Lobby","id":"lobby","privacy":"public","type":"chat","modchat":false,"modjoin":false,"auth":{},"uno":true,"tournaments":"#","hangman":true}``

Comment (panpawn) [inline rooms.js]: why not just send room.game.title, if room.game is a thing

that way, you don't have to send it for each specific game

Comment (Zarel) [inline rooms.js]: Yeah, but, in hindsight, it's a bit much to send to a million users a day when most of them aren't going to use it at all.

Comment (panpawn) [inline rooms.js]: I agree; and plus, bots /could/ use /roomsettings

I mean, I know it's not ideal and they would have to do some parsing, but I'm just saying that all this information with games /is/ already accessible to them.

**Comment by Zarel** [comment]
In hindsight, I actually like `|roomsettings|` as the name for this message.

**Comment by Zarel** [comment]
Replaced by #5279

**Pull request closed without merging**